libsql-server/src/rpc/proxy.rs

```diff
@@ -493,13 +493,18 @@ impl Proxy for ProxyService {
         &self,
         req: tonic::Request<tonic::Streaming<ExecReq>>,
     ) -> Result<tonic::Response<Self::StreamExecStream>, tonic::Status> {
+        let namespace = super::extract_namespace(self.disable_namespaces, &req)?;
+        let namespace_jwt_key = self
+            .namespaces
+            .jwt_key(namespace.clone())
+            .await
+            .map_err(|_| tonic::Status::internal("Error fetching jwt key for a namespace"))?;
         let auth = if let Some(auth) = &self.auth {
-            auth.authenticate_grpc(&req, self.disable_namespaces)?
+            auth.authenticate_grpc(&req, self.disable_namespaces, namespace_jwt_key)?
         } else {
             Authenticated::from_proxy_grpc_request(&req, self.disable_namespaces)?
         };
```

Comment:
I think we can move this all to a fn and make this more abstracted? You don't need to know the T of the request so you can write a fn that accepts a `req: Request<T>`, that should allow us to unify the ones below.

Reply:
When I extracted:
```
+    async fn auth<T>(&self, namespace: NamespaceName, msg: &tonic::Request<T>) -> Result<Authenticated, tonic::Status> {
+        let namespace_jwt_key = self
+            .namespaces
+            .jwt_key(namespace)
+            .await
+            .map_err(|_| tonic::Status::internal("Error fetching jwt key for a namespace"))?;
+        if let Some(auth) = &self.auth {
+            auth.authenticate_grpc(&msg, self.disable_namespaces, namespace_jwt_key)
+        } else {
+            Authenticated::from_proxy_grpc_request(&msg, self.disable_namespaces)
+        }
+    }
```
The compiler complained with:
```
error: future cannot be sent between threads safely
   --> libsql-server/src/rpc/proxy.rs:508:73
    |
508 |       ) -> Result<tonic::Response<Self::StreamExecStream>, tonic::Status> {
    |  _________________________________________________________________________^
509 | |         let namespace = super::extract_namespace(self.disable_namespaces, &req)?;
510 | |         let namespace_jwt_key = self
511 | |             .namespaces
...   |
549 | |         Ok(tonic::Response::new(Box::pin(stream)))
550 | |     }
    | |_____^ future created by async block is not `Send`
    |
    = help: the trait `Sync` is not implemented for `(dyn tonic::codec::Decoder<Item = ExecReq, Error = tonic::Status> + std::marker::Send + 'static)`
note: future is not `Send` as this value is used across an await
   --> libsql-server/src/rpc/proxy.rs:521:55
    |
521 |         let auth = self.auth(namespace.clone(), &req).await?;
    |                                                 ----  ^^^^^ - `&req` is later dropped here
    |                                                 |     |
    |                                                 |     await occurs here, with `&req` maybe used later
    |                                                 has type `&tonic::Request<Streaming<ExecReq>>` which is not `Send`
```
so I ended up just extracting the `namespace_jwt_key()` function.

Reply:
this is quite odd...maybe its some weird interaction with the borrow but I feel like ive done this a million times without issues. But if we are hitting errors we can deal with that later.

Reply:
done